### PR TITLE
[openssl] Fix dynamic build on OSX

### LINKED ISSUE #24089: [openssl] Build error
**Host Environment**
- OS: Mac OS
- Compiler: Xcode 13.1

**To Reproduce**
`./vcpkg install openssl`

**Failure logs**
Looking in `/vcpkg/buildtrees/openssl/install-x64-osx-release-mixed-rel-out.log` it builds ok but at the end I see:
```
FAILED: CMakeFiles/build_libs.util
cd /Users/build/vcpkg/buildtrees/openssl/x64-osx-release-mixed-rel && /usr/bin/install_name_tool -id @rpath/libssl.3.dylib /Users/build/vcpkg/buildtrees/openssl/x64-osx-release-mixed-rel/3.0.2-67d457a95e.clean/libssl.3.dylib && /usr/bin/install_name_tool -id @rpath/libcrypto.3.dylib /Users/build/vcpkg/buildtrees/openssl/x64-osx-release-mixed-rel/3.0.2-67d457a95e.clean/libcrypto.1.1.dylib && /usr/bin/install_name_tool -change /Users/build/vcpkg/packages/openssl_x64-osx-release-mixed/lib/libcrypto.3.dylib @rpath/libcrypto.3.dylib /Users/build/vcpkg/buildtrees/openssl/x64-osx-release-mixed-rel/3.0.2-67d457a95e.clean/libssl.3.dylib
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/install_name_tool: can't open file: /Users/build/vcpkg/buildtrees/openssl/x64-osx-release-mixed-rel/3.0.2-67d457a95e.clean/libcrypto.1.1.dylib (No such file or directory)
```

There is a libcrypto.3.dylib at that location, but not a libcrypto.1.1.dylib 

My triplet file is:
```
set(VCPKG_TARGET_ARCHITECTURE x64)
set(VCPKG_OSX_ARCHITECTURES x86_64)
set(VCPKG_CMAKE_SYSTEM_NAME Darwin)
set(VCPKG_BUILD_TYPE release)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE dynamic)
```

### PR DESCRIPTION
**Pull request by Thomas1664**
**Describe the pull request**

- #### What does your PR fix?
  Fixes #24089

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  Unchanged

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  Yes

Although I can't test it because I don't have a Mac, I'm confident that this should fix the issue.

### CODE REVIEW

**Comment by Thomas1664** [comment]
@winterz Can you please test wether this PR solves your issue?

**Comment by winterz** [comment]
> @winterz Can you please test weither this PR solves your issue?

looks good.

**Comment by winterz** [comment]
can you merge now?

**Comment by Thomas1664** [comment]
> can you merge now?

If @strega-nil-ms has time to review and merge today, it can be merged today, otherwise most likely tomorrow.

**Comment by FrankXie05** [comment]
@Thomas1664 @winterz 
I can't confirm this bug because I don't have the corresponding machine. 😢
I am worried that for the library `libcrypto`, directly changing version to `3.0` will affect the normal use of the 1.1 version.
References: https://www.openssl.org/docs/man3.0/man7/crypto.html

For openssl, the built-in crypto encryption library implements encryption algorithms, but for version maintenance, does it need to be manually changed for each version iteration?
Of course, I think we'll merge this PR, `OpenSSL `upstream pointed out that version 3.0 will be supported until 2026. :)
References: https://www.openssl.org/source/

File: ports/openssl/unix/CMakeLists.txt [hunk 1]
```diff
@@ -263,7 +263,7 @@ if((CMAKE_SYSTEM_NAME STREQUAL "Darwin" OR CMAKE_SYSTEM_NAME STREQUAL "iOS") AND
         COMMAND /usr/bin/install_name_tool -id "${ID_PREFIX}/libssl.${SHLIB_VERSION}.dylib"
                                                "${BUILDDIR}/libssl.${SHLIB_VERSION}.dylib"
         COMMAND /usr/bin/install_name_tool -id "${ID_PREFIX}/libcrypto.${SHLIB_VERSION}.dylib"
-                                               "${BUILDDIR}/libcrypto.1.1.dylib"
+                                               "${BUILDDIR}/libcrypto.3.dylib"
```
Comment (dg0yt) [inline ports/openssl/unix/CMakeLists.txt:266]: Does this have to be hardcoded, or is this also `libcrypto.${SHLIB_VERSION}.dylib`, but in `${BUILDDIR}`?

**Comment by Thomas1664** [comment]
Waiting for #24169 to be merged

**Comment by Thomas1664** [comment]
@FrankXie05 CI is now green

**Review by FrankXie05 — APPROVED** [review]

**Comment by strega-nil-ms** [comment]
Thanks!

**Comment by daschuer** [comment]
This probably does not completely fix the issue. 
I am not able to install libpq see https://github.com/microsoft/vcpkg/issues/22247
Any ideas to fix that?

**Comment by FrankXie05** [comment]
@daschuer 
For cross-compilation, CMake cannot link the correct C standard library normally, so that parameters like `prinft` cannot be passed to the `mian` function implementation definition, so this problem occurs.
Maybe use `CMAKE_C_COMPILER_WORKS` to skip checking, or `CMake with CMAKE_TRY_COMPILE_TARGET_TYPE` to compile static or dynamic libraries instead of `exe` files.
For `community-triplet`, we cannot guarantee that it will build successfully. Feel sorry. 😔

**Comment by daschuer** [comment]
Does this Comment belongs to this PR? 

The issue with building libpq is a native build, using Autotools. The error is 

```
Reason: tried: '/usr/local/lib/libssl.1.1.dylib' (no such file), '/usr/lib/libssl.1.1.dylib' (no such file)
```

**Comment by Thomas1664** [comment]
> Does this Comment belongs to this PR? 
> 
> The issue with building libpq is a native build, using Autotools. The error is 
> 
> ```
> Reason: tried: '/usr/local/lib/libssl.1.1.dylib' (no such file), '/usr/lib/libssl.1.1.dylib' (no such file)
> ```
> 
> Does anyone have an idea to dril that down? 

What is your problem?
Does openssl itself fail to build?
In this case make sure you have the latest catalogue.
Does libpq fail to link openssl?
In this case, I can have a look at it this weekend.

**Comment by daschuer** [comment]
The problem is that libpq builds a test program during the configure step of auto tools. 

This fails my original attempt already contains the patch merged here however I have just merged master again and add a patch that the config.log is preserved https://github.com/microsoft/vcpkg/issues/21422

**Comment by Thomas1664** [comment]
Your issue is not at all related to this PR or the openssl port itself. The issue is that libpq doesn't recognize openssl 3.

**Comment by daschuer** [comment]
I can confirm that the build failure is different. Sorry for the noise.

**Pull request merged**